File: src/lib/sources/google.ts
import * as cheerio from 'cheerio';
import { RawSearchResult } from '@/types';

export async function searchGoogle(keyword: string): Promise<RawSearchResult[]> {
  const searchQuery = keyword.includes(' ') ? `"${keyword}"` : keyword;
  const query = encodeURIComponent(searchQuery);
  const url = `https://www.google.com/search?q=${query}&hl=zh-CN&num=10&tbs=qdr:w`;

  try {
    const res = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml',
        'Accept-Language': 'zh-CN,zh;q=0.9',
        'Accept-Encoding': 'gzip, deflate',
      },
      signal: AbortSignal.timeout(10000),
    });

    if (!res.ok) return [];

    const html = await res.text();
    const $ = cheerio.load(html);
    const results: RawSearchResult[] = [];

    // Google 搜索结果选择器
    $('div.g, div[data-sokoban-container]').each((_, el) => {
      const titleEl = $(el).find('h3').first();
      const linkEl = $(el).find('a[href^="http"]').first();
      const snippetEl = $(el).find('[data-sncf], .VwiC3b, .s3v9rd').first();

      const title = titleEl.text().trim();
      const href = linkEl.attr('href') || '';
      const snippet = snippetEl.text().trim();

      if (title && href && !href.includes('google.com')) {
        results.push({
          title,
          url: href,
          snippet,
          source_type: 'google',
        });
      }
    });

    return results.slice(0, 8);
  } catch (e) {
    console.error('[google] Search failed:', e);
    return [];
  }
}

File: agent-skills/hot-monitor/scripts/sources/google.py
"""Ported from src/lib/sources/google.ts"""

from __future__ import annotations

import httpx
from bs4 import BeautifulSoup
from urllib.parse import quote

from models import RawSearchResult
from sources.bing import UA


async def search_google(keyword: str) -> list[RawSearchResult]:
    search_query = f'"{keyword}"' if " " in keyword else keyword
    url = f"https://www.google.com/search?q={quote(search_query)}&hl=zh-CN&num=10&tbs=qdr:w"

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.get(
                url,
                headers={
                    "User-Agent": UA.replace("120", "124"),
                    "Accept": "text/html,application/xhtml+xml",
                    "Accept-Language": "zh-CN,zh;q=0.9",
                },
            )
        if res.status_code != 200:
            return []

        soup = BeautifulSoup(res.text, "html.parser")
        results: list[RawSearchResult] = []

        for el in soup.select("div.g, div[data-sokoban-container]"):
            title_el = el.select_one("h3")
            link_el = el.select_one('a[href^="http"]')
            snippet_el = el.select_one("[data-sncf], .VwiC3b, .s3v9rd")
            if not title_el or not link_el:
                continue
            title = title_el.get_text(strip=True)
            href = link_el.get("href") or ""
            snippet = snippet_el.get_text(strip=True) if snippet_el else ""
            if title and href and "google.com" not in href:
                results.append(
                    {
                        "title": title,
                        "url": href,
                        "snippet": snippet,
                        "source_type": "google",
                    }
                )

        return results[:8]
    except Exception as e:
        print(f"[google] Search failed: {e}", flush=True)
        return []
